### ISSUE TITLE
AWS SDK v2

### ISSUE THREAD

**Issue by darthShadow** [rclone/rclone#4989]
#### What is your current rclone version (output from `rclone version`)?

N/A

#### What problem are you are trying to solve?

Upgrade AWS S3 SDK to v2

#### How do you think rclone should be changed to solve that?

Upgrade AWS S3 SDK to v2

Relevant General Availability Blog from AWS containing the list of improvements:

https://aws.amazon.com/blogs/developer/aws-sdk-for-go-version-2-general-availability/

Note: This will require Go 1.15+ so should probably be targeted once we have that as the minimum version for rclone.

**Comment by ncw** [comment]
Thanks for making an issue about this.

I suspect it might be quite a lot of work changing to the v2 API alas.

We could do the work now but stop the s3 backend compiling on anything less than go1.15.

**Comment by darthShadow** [comment]
That's fine. Just wanted to put this on your radar since v1 will eventually get deprecated now that v2 has been promoted to GA and there are some nice QoL improvements too.

**Comment by ncw** [comment]
Good idea!

**Comment by tareqhs** [comment]
hi, is anyone working on this?

**Comment by ncw** [comment]
@tareqhs not at the moment.  Is it something you are interested in helping with? What advantages do you think we'd get from the new API?

I dread having to port the service configuration from v1 to v2 as we don't have many tests and there are 100s of different ways of configuring it!

**Comment by tareqhs** [comment]
@ncw I'll count the benefits in an opinionated descending order:
1. v2 is taking over and v1 will be deprecated sooner or later. for corporate usage this is a significant factor.
1. performance improvement (claimed) - this is questionable and depends on the target scenarios.
1. the huge sdk has been broken to many submodules which is great for rclone because it uses the S3 submodule only.
1. the APIs have been modernized for go 1.15 (e.g. context usage) which aligns better with rclone's standards.

the configuration flow seems actually simpler in v2. I haven't performed a deep dive for rclone's use case though...

made a very small step in my own branch, probably will take time to finish so cooperation is welcome
https://github.com/tareqhs/rclone/tree/aws_sdk_v2

**Comment by ncw** [comment]
All good reasons, and it is generally our policy to keep up with dependencies.

Your branch doesn't look too scary at the moment. Does it work at all?

Does the V2 sdk let you call putobject without giving it an io.Seeker? Rclone only has io.Readers so that would be a major improvement if it does and would mean we could get rid of the presigned request which isn't very compatible with other S3 implementations (snowball doesn't support it).

**Comment by tareqhs** [comment]
@ncw Yes, v2 accepts a minimal `io.Reader` for `PutObject`:
https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/s3@v1.19.1#PutObjectInput

Since the change is big and affects a popular backend, I suggest we first reach parity with existing code and then start to refactor to make better use of v2's new facilities. Still I made a separate commit that replaces the bulky signed request with a simple direct usage of `io.Reader`.

I still don't have a fully compiled code but I've already migrated everything except for s3 client configuration, which should be that complicated. (I'll come back to this ASAP)

How can test this? Are there any test suites for rclone+s3 ?

**Comment by ncw** [comment]
> @ncw Yes, v2 accepts a minimal `io.Reader` for `PutObject`: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/s3@v1.19.1#PutObjectInput

That's great!
 
> Since the change is big and affects a popular backend, I suggest we first reach parity with existing code and then start to refactor to make better use of v2's new facilities. Still I made a separate commit that replaces the bulky signed request with a simple direct usage of `io.Reader`.

That sounds like a good plan.

We could also temporarily fork the s3 backend into an s3v2 backend and have both in the codebase for a while so we could get more testing like that. Or we can make a branch and issue betas from the branch very easily.
 
> I still don't have a fully compiled code but I've already migrated everything except for s3 client configuration, which should be that complicated. (I'll come back to this ASAP)

OK

> How can test this? Are there any test suites for rclone+s3 ?

The best way to test is using [the integration tests](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#backend-testing). In brief you can run `go test` while in the `backend/s3` directory provided you've set up a `TestS3:` remote. This tests nearly all the backend features, so iterate until this passes. (Or use your existing remote with `go test -remote mys3remote:`)

To run all the integration tests you would then (from the source root)

    go install ./...
    test_all -remotes TestS3:

Which will run rclone's integration tests for sync/operations/vfs/backend/etc against `TestS3:` which is the comprehensive test! The integration test machine is configured to run this against quite a few different s3 providers.

This doesn't test the incredibly varied matrix of client configurations though - there aren't unit tests for those.

**Comment by darthShadow** [comment]
@ncw Looks like v1 will be going into Maintenance mode in July '24 & EOL in July '25: https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-sdk-for-go-v1-on-july-31-2025/

**Comment by ncw** [comment]
We'd better get on with this :-) @tareksha made a start on this in https://github.com/tareksha/rclone/tree/aws_sdk_v2 - I doubt we'll be able to rebase that easily on master but it should give some pointers.

I will do this in v1.67 release unless I get any other takers before then? Maybe you'd like to continue @tareksha ?

**Comment by ncw** [comment]
I am in the process of doing this for 1.68 and as soon as I have something working, I will post a beta.

@wiserain I note that the pikpak backend uses the aws s3 v1 sdk also so it will need updating too. I think that should be quite easy though as it is very lightly used.

**Comment by ncw** [comment]
Here is the first version of the s3 backend using the v2 SDK

[v1.68.0-beta.8133.ba28a7e64.fix-4989-s3-aws-sdk-v2](https://beta.rclone.org/branch/fix-4989-s3-aws-sdk-v2/v1.68.0-beta.8133.ba28a7e64.fix-4989-s3-aws-sdk-v2/) on branch [fix-4989-s3-aws-sdk-v2](https://github.com/rclone/rclone/tree/fix-4989-s3-aws-sdk-v2) 

Here is the status so far

Changes
    
- `--s3-sts-endpoint` is no longer supported
- `--s3-use-unsigned-payload` to control use of trailer checksums
    
Not implemented yet
    
- `--s3-use-presigned-request` - does anyone actually use this? Tempted to deprecate it.
    
Untested
    
- v2auth
- fancy auth in ec2, kubernetes, profiles
    
I am particularly interested in testing with the more esoteric AWS authentication methods. AWS s3 has a massive matrix for auth methods and no way I can test them all. I think they should work - the SDK has been improved massively here, but they may work differently to they used to.

The integration tests all pass with AWS and Minio - I haven't tested any other providers yet.

This is a big change:

    8 files changed, 549 insertions(+), 526 deletions(-)

So any and all testing appreciated. Please report problems here.

**Comment by ncw** [comment]
@wiserain let me know if you need help. I've learned the new SDK now!

**Comment by ncw** [comment]
Thanks @wiserain - I'll cherry-pick that into my branch and fix the go.mod conflicts.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.68 

Any testing appreciated.

Please comment here or make a new issue if you have a problem.

**Comment by jahands** [comment]
Tested with Cloudflare R2 on MacOS - everything seems to work fine, but copying between two R2 buckets no longer shows progress bars. Is that expected?

**Comment by jahands** [comment]
Confirmed:
```
open chunk writer: started multipart upload:
```

**Comment by ncw** [comment]
Thanks. Can you give me the command line used and the size of files copied and I'll try to replicate. Thanks @jahands

**Comment by darthShadow** [comment]
@ncw I am able to replicate the same behaviour for every transfer. This seems to be happening only for multi-part copies, and the stats are completely zero for the file (in terms of the data & speed) even after the transfer is done. This is using `cp` to copy a file into an `rclone mount` of an S3-backed Crypt remote.

Let me know if you are unable to reproduce this and I will share further details.

**Comment by ncw** [comment]
@darthShadow I tried rclone copy the file and mounting and cp -a  and with --vfs-cache-mode off and --vfs-cache-mode writes and everything looked normal with an s3 backeed (minio) crypt remote.

Please share more details - thank you :-)

**Comment by darthShadow** [comment]
> Please share more details

Sent a message on the forum with the config file & debug logs. FWIW, I did run a `git bisect` too, just to be sure, and it came to the same commit:
https://github.com/rclone/rclone/commit/fd9049c83d9ef181170d18467b16f35fc5f11599

**Comment by ncw** [comment]
With a bit of help from @darthShadow and a bit of staring at the code and looking at traces I managed to figure out the problem.

This fixes any servers with the no accounting problem but causes double accounting on other servers :frowning: 

```
diff --git a/backend/s3/s3.go b/backend/s3/s3.go
index 2ca9b4fb3..90b963122 100644
--- a/backend/s3/s3.go
+++ b/backend/s3/s3.go
@@ -5988,7 +5988,7 @@ func (w *s3ChunkWriter) WriteChunk(ctx context.Context, chunkNumber int, reader
 	if do, ok := reader.(pool.DelayAccountinger); ok {
 		// To figure out this number, do a transfer and if the accounted size is 0 or a
 		// multiple of what it should be, increase or decrease this number.
-		do.DelayAccounting(3)
+		do.DelayAccounting(2)
 	}
 
 	// create checksum of buffer for integrity checking
```

The difference seems to be that chunks are sent as UNSIGNED-PAYLOAD on troublesome servers but as signed chunks on Minio - so the Minio is reading the data one more time to calculate the signature hash.

I don't know why that differs between the servers - that is a mystery. They chunks should be signed on both servers.

That must be down to the SDK as the config is identical between servers.

This replicates perfectly with R2

I'll see if I can come up with a fix, but I'm going to need to find out why some chunks are sent with UNSIGNED-PAYLOAD and that will require digging in the SDK source I think.

**Comment by darthShadow** [comment]
@ncw Correct me, if I am wrong, but don't we set it to use Unsigned Payloads ourselves here:

https://github.com/rclone/rclone/blob/master/backend/s3/s3.go#L6127

It's set to `false` only for AWS, which implies that the bug is actually with Minio and the others are behaving correctly.

**Comment by darthShadow** [comment]
I think this explains the difference with v1 for the hashing:

https://github.com/aws/aws-sdk-go-v2/issues/1040#issuecomment-1002722159

```
While investigating this it looks like the v1 SDK had a [customiation](https://github.com/aws/aws-sdk-go/blob/136908423548340fc3b7dbce972ba14b79942ffa/service/s3/body_hash.go#L27-L30) papering over this issue by always computing the UploadPart's ContentMD5 parameter. V1 SDK still ignored the ContentMD5 input parameter for multi-part, but would add its own ContentMD5 under the hood.
```

**Comment by darthShadow** [comment]
It also looks like Minio added support for signed trailers more than a year ago:

https://github.com/minio/minio/pull/16484

https://github.com/minio/minio/releases/tag/RELEASE.2023-05-18T00-05-36Z

**Comment by ncw** [comment]
@darthShadow 
> Correct me, if I am wrong, but don't we set it to use Unsigned Payloads ourselves here:
> 
> https://github.com/rclone/rclone/blob/af9f8ced805509c3efd264fc5be2a82ea1a000f2/backend/s3/s3.go#L6127-L6131
> 
> It's set to `false` only for AWS, which implies that the bug is actually with Minio and the others are behaving correctly.

I think that is only for single part uploads. The multipart uploads shouldn't be affected.

> I think this explains the difference with v1 for the hashing and the accounting delay count reduction:
> 
> [aws/aws-sdk-go-v2#1040 (comment)](https://github.com/aws/aws-sdk-go-v2/issues/1040#issuecomment-1002722159)

That is a very interesting thread. We don't use the upload manager as rclone has a framework internally to do the same thing but I wonder if the logic for it went into the UploadPart call - it would make sense if it did.

Note also that the signature is X-Amz-Content-Sha256 which is set to UNSIGNED-PAYLOAD for R2 and it shouldn't be.

> > While investigating this it looks like the v1 SDK had a [customiation](https://github.com/aws/aws-sdk-go/blob/136908423548340fc3b7dbce972ba14b79942ffa/service/s3/body_hash.go#L27-L30) papering over this issue by always computing the UploadPart's ContentMD5 parameter. V1 SDK still ignored the ContentMD5 input parameter for multi-part, but would add its own ContentMD5 under the hood.

We do calculate the ContentMD5 and set it - maybe we don't need to any more? I tried that and it didn't seem to make any difference.

> It also looks like Minio added support for signed trailers more than a year ago:
> 
> [minio/minio#16484](https://github.com/minio/minio/pull/16484)
> 
> https://github.com/minio/minio/releases/tag/RELEASE.2023-05-18T00-05-36Z

Hmm, interesting. I don't know how to enable that in the SDK though. It is probably relevant to this though.

```
$ rclone copy -vv --s3-use-unsigned-payload=false /tmp/10M TestS3MinioManual:rclone/z/ --dump headers
2024/08/30 13:37:31 Failed to copy: operation error S3: PutObject, failed to compute payload hash: failed to seek body to start, request stream is not seekable
```

Doesn't seem to work in the docker image I have yet though

**Comment by darthShadow** [comment]
> It is probably relevant to this though.

Any chance you are accessing Minio via HTTP instead of HTTPS?

Because the s3 source seems to mandate signing only for HTTP, not HTTPS.

From the source (https://github.com/aws/aws-sdk-go-v2/blob/main/service/internal/checksum/middleware_compute_input_checksum.go#L188-L191):

```
// If trailing checksums are not enabled but protocol is still HTTPS
// disabling computing the payload hash. Downstream middleware  handler
// (ComputetPayloadHash) will set the payload hash to unsigned payload,
// if signing was used.
```

**Comment by ncw** [comment]
Yes I think you are right @darthShadow it is minion accessed over http that is the odd one out in working. Everything else including aws is broken.

Should be easy to fix now.

**Comment by ncw** [comment]
Can you give this a go @darthShadow and @jahands ? I'm hoping this will fix the accounting issue.

[v1.68.0-beta.8250.1bf4819e0.fix-4989-s3-accounting](https://beta.rclone.org/branch/fix-4989-s3-accounting/v1.68.0-beta.8250.1bf4819e0.fix-4989-s3-accounting/) on branch [fix-4989-s3-accounting](https://github.com/rclone/rclone/tree/fix-4989-s3-accounting) (uploaded in 15-30 mins)

**Comment by darthShadow** [comment]
All good from my end, thanks.

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.68 

I'll close this issue again in the hope that that really was the last problem!
